workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/utilities/TokenBuilder.kt

```diff
@@ -0,0 +1,171 @@
+     *
+     * @param callbackSetter A function that sets a particular member property
+     */
+    private fun initializeAmountOrThrow(callbackSetter: () -> Unit): TokenBuilder {
+        if (listOf(longAmount, intAmount, doubleAmount, bigDecimalAmount).all { it == null }) {
+            callbackSetter()
+        } else {
+            throw TokenBuilderException("The token amount has already been initialized")
+        }
+        return this
+    }
+
+    /**
+     * Gets the amount value that has been initialized.
+     */
+    private fun getInitializedAmount() = listOfNotNull(longAmount, intAmount, doubleAmount, bigDecimalAmount).single()
+
+    /**
+     * Allow the developer to add an amount in any of the following classes to the TokenBuilder
+     * [Long], [Int], [Double], [BigDecimal]. The types of the amount values must be preserved
+     * as their eventual conversion to [Long] requires information from the [TokenType] which
+     * may or may not be initialized yet.
+     */
+
+    /**
+     * Initialize the [longAmount] member property of the builder. Throws an exception if another
+     * possible amount has already been initialized.
+     *
+     * @param longAmount The [Long] we will use to set the [longAmount] member property.
+     */
+    fun withAmount(longAmount: Long) = initializeAmountOrThrow { this.longAmount = longAmount }
+
+    /**
+     * Initialize the [intAmount] member property of the builder. Throws an exception if another
+     * possible amount has already been initialized.
+     *
+     * @param intAmount The [Int] we will use to set the [intAmount] member property.
+     */
+    fun withAmount(intAmount: Int) = initializeAmountOrThrow { this.intAmount = intAmount }
+
+    /**
+     * Initialize the [doubleAmount] member property of the builder. Throws an exception if another
+     * possible amount has already been initialized.
+     *
+     * @param doubleAmount The [Double] we will use to set the [doubleAmount] member property.
+     */
+    fun withAmount(doubleAmount: Double) = initializeAmountOrThrow { this.doubleAmount = doubleAmount }
+
+    /**
+     * Initialize the [bigDecimalAmount] member property of the builder. Throws an exception if another
+     * possible amount has already been initialized.
+     *
+     * @param bigDecimalAmount The [BigDecimal] we will use to set the [bigDecimalAmount] member property.
+     */
+    fun withAmount(bigDecimalAmount: BigDecimal) = initializeAmountOrThrow { this.bigDecimalAmount = bigDecimalAmount }
+
+    /**
+     * Replicates the Kotlin DSL [of] infix function. After supplying a [TokenType] to the builder
+     * the developer will be able to call [resolveAmountTokenType] to access the [amountTokenType]
+     * member property.
+     *
+     * @param t The token type that will be used to initialize the [amountTokenType] member property.
+     */
+    fun <T: TokenType> of(t: T): TokenBuilder = when (getInitializedAmount()) {
```

Comment:
The builder should be converting itself to a value only when you call `build() function, not when you supply values to it.

Reply:
Makes sense! Fixed this.